modules/core/src/main/scala/com.snowplowanalytics.snowplow.bigquery/processing/TableManager.scala

```diff
@@ -44,75 +47,106 @@ object TableManager {
 
   private implicit def logger[F[_]: Sync] = Slf4jLogger.getLogger[F]
 
+  trait WithHandledErrors[F[_]] {
+    def addColumns(columns: Vector[Field]): F[Unit]
+    def createTable: F[Unit]
+  }
+
   def make[F[_]: Async](
     config: Config.BigQuery,
-    retries: Config.Retries,
-    credentials: Credentials,
-    appHealth: AppHealth[F],
-    monitoring: Monitoring[F]
+    credentials: Credentials
   ): F[TableManager[F]] =
     for {
       client <- Sync[F].delay(BigQueryOptions.newBuilder.setCredentials(credentials).build.getService)
-    } yield impl(config, retries, client, appHealth, monitoring)
+    } yield impl(config, client)
 
-  private def impl[F[_]: Async](
-    config: Config.BigQuery,
+  def withHandledErrors[F[_]: Async](
+    underlying: TableManager[F],
     retries: Config.Retries,
-    client: BigQuery,
     appHealth: AppHealth[F],
     monitoring: Monitoring[F]
+  ): F[WithHandledErrors[F]] =
+    for {
+      addingColumnsEnabled <- Ref[F].of[Boolean](true)
+    } yield new WithHandledErrors[F] {
+      def addColumns(columns: Vector[Field]): F[Unit] =
+        addingColumnsEnabled.get.flatMap {
+          case true =>
+            BigQueryRetrying
+              .withRetries(appHealth, retries, monitoring, Alert.FailedToAddColumns(columns.map(_.name), _)) {
+                Logger[F].info(s"Altering table to add columns [${showColumns(columns)}]") *>
+                  underlying
+                    .addColumns(columns)
+                    .recoverWith(handleTooManyColumns(retries, monitoring, addingColumnsEnabled, columns))
+                    .onError(logOnRaceCondition)
+              }
+          case false =>
+            Async[F].unit
+        }
+
+      def createTable: F[Unit] =
+        BigQueryRetrying.withRetries(appHealth, retries, monitoring, Alert.FailedToCreateEventsTable(_)) {
+          underlying.createTable
+        }
+    }
+
+  private def impl[F[_]: Async](
+    config: Config.BigQuery,
+    client: BigQuery
```

Comment:
Another idea...how hard would be to mock `BigQuery` client? Then we could test whole our code!

Reply:
I thought about this... so the trait would be something like

```scala
trait BigQueryClient[F[_]] {
  def getTable(dataset: String, table: String): F[Table]
}
```

But unfortunately that cannot be mocked in a unit test, because [there are no public constructors for Table or its builder](https://github.com/googleapis/java-bigquery/blob/a6402b16ffa8ab473c415cf87a6b2e0d8068e3dc/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/Table.java#L39).  So then I need to mock `Table` as well....

I would end up mocking more and more, and it wouldn't really increase the amount of real code that gets tested.